Make stdout unbuffered

kt3k opened the pull request:
This PR fixes #948. Continued from https://github.com/ry/deno/commit/756078733185e9127901c6957a73d0e2c9bdafde.

I copied `//js/testing/*` to `//tools/ts_builder_library/testing/*` (and slightly modified them to work with node.js) because `//js/testing` is not isomorphic anymore (doesn't work in (ts-)node because of `import ... from "deno"` line) and use it in ts_builder_library testing.

ry commented:
I've been hesitating on doing this because I think it's nice if we can maintain browser compatibility with the testing library.. It's such a minor thing to loose compatibility for.

One idea is to abuse the `console.group()` API for printing to the same line. Something like this:
```js
console.group()
console.log("testing foobar....")
// ....
console.log("OK")
console.groupEnd()
``` 
Although printing to the same line isn't exactly what `console.group` is for, it's somewhat similar... It would allow us to still use `testing` in the browser.

kt3k commented:
@ry OK. I'll separate this PR into 2 pieces. line buffer related changes and testing library changes.

justjavac commented:
I think we should not use `group`, maybe we can choose `groupCollapsed` instead. and `group` can also be compatible with other platforms, such as Node.js/Browsers.

kt3k commented:
- Limitted this PR to stdout changes.
- Added a test case for checking output without line break. (//tests/stdout.ts)

I still don't figure out how to create unbuffered stdout in windows.

ry commented on src/resources.rs:
```diff
@@ -43,14 +43,34 @@ pub type ResourceId = u32; // Sometimes referred to RID.
+    let stdout = unsafe {std::fs::File::from_raw_handle(kernel32::GetStdHandle(winapi::winbase::STD_OUTPUT_HANDLE))};
+
+    m.insert(1, Repr::Stdout(
+        tokio::fs::File::from_std(stdout)));
```
consider rewriting like this
```

    m.insert(1, Repr::Stdout({
      #[cfg(not(windows))]
      let stdout = unsafe {std::fs::File::from_raw_fd(1)};
      #[cfg(windows)]
      let stdout = unsafe { std::fs::File::from_raw_handle(kernel32::GetStdHandle(winapi::winbase::STD_OUTPUT_HANDLE))};
        tokio::fs::File::from_std(stdout)
  }));

```

kt3k commented:
Now build and tests pass both in travis and appveyor. Please review.

ry commented:
@kt3k Cool - looks good - sorry for the delay. Can you please apply the following clean ups:

```patch
From 62cd858fe927498c171e0954677a6beab5ceb0fd Mon Sep 17 00:00:00 2001
From: Ryan Dahl <ry@tinyclouds.org>
Date: Fri, 21 Dec 2018 12:53:59 -0500
Subject: [PATCH] renaming, formatting

---
 src/resources.rs                                   | 9 +++++++--
 tests/stdout.test                                  | 2 --
 tests/unbufferred_stdout.test                      | 2 ++
 tests/{stdout.ts => unbufferred_stdout.ts}         | 0
 tests/{stdout.ts.out => unbufferred_stdout.ts.out} | 0
 5 files changed, 9 insertions(+), 4 deletions(-)
 delete mode 100644 tests/stdout.test
 create mode 100644 tests/unbufferred_stdout.test
 rename tests/{stdout.ts => unbufferred_stdout.ts} (100%)
 rename tests/{stdout.ts.out => unbufferred_stdout.ts.out} (100%)

diff --git a/src/resources.rs b/src/resources.rs
index a69cae3..64ec959 100644
--- a/src/resources.rs
+++ b/src/resources.rs
@@ -65,13 +65,18 @@ lazy_static! {
 
     m.insert(1, Repr::Stdout({
       #[cfg(not(windows))]
-      let stdout = unsafe {std::fs::File::from_raw_fd(1)};
+      let stdout = unsafe { std::fs::File::from_raw_fd(1) };
       #[cfg(windows)]
-      let stdout = unsafe {std::fs::File::from_raw_handle(kernel32::GetStdHandle(winapi::um::winbase::STD_OUTPUT_HANDLE))};
+      let stdout = unsafe {
+        std::fs::File::from_raw_handle(kernel32::GetStdHandle(
+            winapi::um::winbase::STD_OUTPUT_HANDLE))
+      };
       tokio::fs::File::from_std(stdout)
     }));
 
+    // TODO Should stderr also be unbufferred?
     m.insert(2, Repr::Stderr(tokio::io::stderr()));
+
     m
   });
 }
diff --git a/tests/stdout.test b/tests/stdout.test
deleted file mode 100644
index c66f5c2..0000000
--- a/tests/stdout.test
+++ /dev/null
@@ -1,2 +0,0 @@
-args: tests/stdout.ts --reload
-output: tests/stdout.ts.out
diff --git a/tests/unbufferred_stdout.test b/tests/unbufferred_stdout.test
new file mode 100644
index 0000000..94475cf
--- /dev/null
+++ b/tests/unbufferred_stdout.test
@@ -0,0 +1,2 @@
+args: tests/unbufferred_stdout.ts --reload
+output: tests/unbufferred_stdout.ts.out
diff --git a/tests/stdout.ts b/tests/unbufferred_stdout.ts
similarity index 100%
rename from tests/stdout.ts
rename to tests/unbufferred_stdout.ts
diff --git a/tests/stdout.ts.out b/tests/unbufferred_stdout.ts.out
similarity index 100%
rename from tests/stdout.ts.out
rename to tests/unbufferred_stdout.ts.out
-- 
2.15.0

```

ry commented:
@kt3k I think we ought to also make stderr unbuffered now as well.

kt3k commented:
Updated and squashed the commits.

After checking stderr in some versions of deno, it seems already unbuffered. I only added test of `unbuffered_stderr` and it seems passing in travis and appveyor without change of impl.
